olive/passes/onnx/moe_experts_distributor.py

```diff
@@ -0,0 +1,259 @@
+            if prefix_a[i] == "/":
+                last_slash = i
+        else:
+            break
+
+    if last_slash > 0:
+        prefix = prefix[: last_slash + 1]
+
+    if not prefix.endswith("/"):
+        prefix += "/"
+    prefix += op_type
+
+    return model.create_node_name(op_type, prefix)
+
+
+def _create_ranked_model(
+    nodes: Dict[str, Message],
+    producers: Dict[str, Message],
+    consumers: Dict[str, Message],
+    experts: List[List[str]],
+    world_size: int,
+    num_experts: int,
+    rank: int,
+):
+    experts_per_rank = int(num_experts / world_size)
+    start_index = rank * experts_per_rank
+    end_index = start_index + experts_per_rank
+
+    for expert in experts:
+        div_node = nodes[expert[5]]
+        concat_node = nodes[expert[13]]
+        reshape_node = nodes[expert[1]]
+
+        div_node_oname = div_node.output[0]
+        reshape_node_oname = reshape_node.output[0]
+
+        mul_nodes = [n for n in consumers[div_node_oname] if n.op_type == "Mul"]
+        matmul_nodes = [producers[iname] for iname in concat_node.input if producers[iname].op_type == "MatMul"]
+        slice_nodes = [n for n in consumers[reshape_node_oname] if n.op_type == "Slice"]
+
+        slice_nodes[start_index].input[1] = slice_nodes[0].input[1]
+
+        for i in range(num_experts):
+            if (i < start_index) or (i >= end_index):
+                mul_nodes[i].input.remove(div_node_oname)
+                concat_node.input.remove(matmul_nodes[i].output[0])
+
+                for iname in list(slice_nodes[i].input):
+                    slice_nodes[i].input.remove(iname)
+
+
+def _insert_commop_nodes(model: OnnxModel, nodes: Dict[str, Message], experts: List[List[str]], world_size: int):
+    for expert in experts:
+        for i, j in [(0, 1), (-2, -1)]:
+            prev_node = nodes[expert[i]]
+            next_node = nodes[expert[j]]
+
+            node_name = _create_node_name(model, "AllToAll", prev_node.name, next_node.name)
+            output_name = node_name + "_output_0"
+            alltoall = onnx.helper.make_node(
+                "AllToAll", prev_node.output, [output_name], name=node_name, domain=_domain, group_size=world_size
+            )
+
+            model.add_node(alltoall)
+            OnnxModel.replace_node_input(next_node, prev_node.output[0], output_name)
+
+
+def _replace_constant_value(constant: Message, value: int):
+    for attr in constant.attribute:
+        if attr.name == "value":
+            attr.CopyFrom(
+                onnx.helper.make_attribute(
+                    attr.name, onnx.numpy_helper.from_array(numpy.array([value], numpy.int64)), attr.doc_string
+                )
+            )
+            return
+
+
+def _fix_shapes(
+    model: OnnxModel,
+    nodes: Dict[str, Message],
+    producers: Dict[str, Message],
+    consumers: Dict[str, Message],
+    experts: List[List[str]],
+    world_size: int,
+    num_experts: int,
+    rank: int,
```

Comment:
`rank` is not used in this function, do we  need to keep it here still? Maybe any other case will use `rank` in the future?

Reply:
Removed!